### PR TITLE
fixed arguments in mul_poly

### PR DESCRIPTION
**Pull request by srajangarg**
When we need to use the `RCP` of one of the arguments of the function, ( as required here) we cannot pass the reference to the class itself. This is because, we'll construct a stack variable, containing the required RCP of the argument. When the freeing of stack happens on the return of the function, it throws a `free() invalid pointer` error. As that memory location was not allocated on the heap.

This is also indicated by what certik said in https://github.com/symengine/symengine/pull/406#issuecomment-77592690

### CODE REVIEW

**Comment by Sumith1896** [comment]
+1 if tests pass.

File: symengine/polynomial.cpp [hunk 1]
```diff
@@ -250,23 +250,23 @@ unsigned int bit_length(T t){
     return count;
 }
 
-RCP<const UnivariateIntPolynomial> mul_poly(RCP<const UnivariateIntPolynomial> a, RCP<const UnivariateIntPolynomial> b) {
-    //TODO: Use `const RCP<const UnivariateIntPolynomial> &a` for input arguments,
-    //      even better is use `const UnivariateIntPolynomial &a`
```
Comment (isuruf) [inline symengine/polynomial.cpp]: > even better is use `const UnivariateIntPolynomial &a`

Comment (srajangarg) [inline symengine/polynomial.cpp]: It's not possible to do so. Updating description.

**Comment by Sumith1896** [comment]
@isuruf Views?

**Comment by isuruf** [comment]
I still don't get why `RCP` is needed. Where is it needed?

**Comment by srajangarg** [comment]
You were right, I have managed to make the arguments `UnivariateIntPolynomial&`. Initially, the function was constructing a new negative polynomial on the heap (if it was required). This was redundant. I just had to add support for `max_coeff` handling the case if the whole polynomial was negative of what it currently is. This was done using a new parameter `negative` which defaults to `false`.

File: symengine/polynomial.cpp [hunk 2]
```diff
@@ -130,11 +130,23 @@ vec_basic UnivariateIntPolynomial::get_args() const {
     return args;
 }
 
-integer_class UnivariateIntPolynomial::max_coef() const {
-    integer_class curr = dict_.begin()->second;
+integer_class UnivariateIntPolynomial::max_coef(bool negative) const {
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Shouldn't this give the maximum absolute value coefficient?

File: symengine/polynomial.cpp [hunk 3]
```diff
@@ -274,8 +282,8 @@ RCP<const UnivariateIntPolynomial> mul_poly(RCP<const UnivariateIntPolynomial> a
     a1 <<= N;
     integer_class a2 = a1 / 2;
     integer_class mask = a1 - 1;
-    integer_class sa = a->eval_bit(N);
-    integer_class sb = b->eval_bit(N);
+    integer_class sa = a_sign * a.eval_bit(N);
```
Comment (isuruf) [inline symengine/polynomial.cpp]: `a_sign` is not really needed, right?

Comment (srajangarg) [inline symengine/polynomial.cpp]: It is, according to how the old code was written.

Comment (isuruf) [inline symengine/polynomial.cpp]: Yes, but I don't think it is needed.

File: symengine/polynomial.cpp (hunk 2 above)
Comment (srajangarg) [inline symengine/polynomial.cpp]: @Sumith1896 wrote it initially. But I believe it should be absolute value.

File: symengine/polynomial.cpp (hunk 3 above)
Comment (srajangarg) [inline symengine/polynomial.cpp]: Tests don't pass when I tried changing it. I think it requires the highest degree term to turn out positive maybe?

**Comment by Sumith1896** [comment]
Could you resolve the merge conflicts? The Polynomial PR was merged.

**Comment by srajangarg** [comment]
Sure

**Comment by certik** [comment]
Please do not mix styling changes with fixing bugs or implementing new features. It's really hard for me to review it, since most of the changes just move `{` around.

We should start using #866, then at least the formatting changes should go away.

**Comment by srajangarg** [comment]
Sorry @certik, will keep in mind from now on.

**Comment by srajangarg** [comment]
@certik @isuruf Does this require any work?

File: symengine/polynomial.cpp [hunk 4]
```diff
@@ -174,19 +182,31 @@ integer_class UnivariateIntPolynomial::eval_bit(const int &x) const {
     return result;
 }
 
-bool UnivariateIntPolynomial::is_zero() const {
-    return dict_.empty();
+bool UnivariateIntPolynomial::is_zero() const
+{
+    if (dict_.empty())
+        return true;
+    return false;
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Why this change?

**Comment by isuruf** [comment]
@srajangarg, can you remove the styling changes?

**Comment by srajangarg** [comment]
@isuruf done. Sorry about another commit, I had merged conflicts and made style changes together in the previous commit, so couldn't reset it.

**Comment by Sumith1896** [comment]
@isuruf How does it look to you? Shall I merge this?

**Comment by isuruf** [comment]
@srajangarg, can you add test cases where there are negative coefficients and the largest absolute value coefficient is the constant term?

**Comment by srajangarg** [comment]
@isuruf There was a huge mistake in `mul_poly`. The bound for N was incoorect!
The actual bound can be seen [here](http://www.cs.berkeley.edu/~fateman/papers/polysbyGMP.pdf#page=3) (page 3, para 1). I have corrected this.

I also added some test cases.

File: symengine/tests/basic/test_polynomial.cpp [hunk 5]
```diff
@@ -76,14 +76,26 @@ TEST_CASE("Subtracting two UnivariateIntPolynomial", "[UnivariateIntPolynomial]"
 TEST_CASE("Multiplication of two UnivariateIntPolynomial", "[UnivariateIntPolynomial]")
 {
     RCP<const Symbol> x  = symbol("x");
-    RCP<const UnivariateIntPolynomial> a = univariate_int_polynomial(x, {{0, 1_z}, {1, 2_z}, {2, 1_z}});
-    RCP<const UnivariateIntPolynomial> b = univariate_int_polynomial(x, {{0, -1_z}, {1, -2_z}, {2, -1_z}});
+    map_uint_mpz adict_ = {{0, 1_z}, {1, 2_z}, {2, 1_z}};
+    map_uint_mpz bdict_ = {{0, -1_z}, {1, -2_z}, {2, -1_z}};
```
Comment (isuruf) [inline symengine/tests/basic/test_polynomial.cpp]: Can you add a test case like `{{0, -1},{1,-2},{2, -100}}`?

**Comment by isuruf** [comment]
Thanks for looking into it. `mul_poly` needs lots more testing.

**Comment by CodeMaxx** [comment]
@Sumith1896 I'm working on improving coverage for `polynomial.cpp` and some other files as well.

**Comment by srajangarg** [comment]
@isuruf ping

File: symengine/polynomial.cpp [hunk 6]
```diff
@@ -190,14 +190,16 @@ integer_class UnivariateIntPolynomial::eval_bit(const int &x) const
 {
 
     unsigned int last_deg = dict_.rbegin()->first;
-    integer_class result(0);
+    integer_class result(0), two_pow;
 
     for (auto it = dict_.rbegin(); it != dict_.rend(); ++it) {
 
-        result = (*it).second + (result << x * (last_deg - (*it).first));
+        two_pow = integer_class(1) << x * (last_deg - (*it).first);
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Why this change?

File: symengine/polynomial.cpp [hunk 7]
```diff
@@ -190,14 +190,16 @@ integer_class UnivariateIntPolynomial::eval_bit(const int &x) const
...
         last_deg = (*it).first;
+        result = (*it).second + two_pow * result;
     }
-    result = result << x * last_deg;
+    two_pow = integer_class(1) << x * last_deg;
+    result *= two_pow;
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Why this change?

File: symengine/polynomial.cpp [hunk 8]
```diff
@@ -314,70 +316,48 @@ unsigned int bit_length(T t)
...
-    RCP<const Symbol> var = symbol("");
-    if (a->get_var()->get_name() == "") {
-        var = b->get_var();
-    } else if (b->get_var()->get_name() == "") {
-        var = a->get_var();
-    } else if (!(a->get_var()->__eq__(*b->get_var()))) {
-        throw std::runtime_error("Error: variables must agree.");
-    } else {
-        var = a->get_var();
-    }
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Why this change?

**Comment by isuruf** [comment]
I think something went wrong while resolving conflicts. Can you check?

File: symengine/polynomial.cpp (hunk 8 above)
Comment (srajangarg) [inline symengine/polynomial.cpp]: This wasn't needed. Just need to check if the variables are equal or not, which I have added below.

Comment (isuruf) [inline symengine/polynomial.cpp]: It is needed. There can be polynomials with no var, just a coefficient which can be used later.
I think we need to rethink the design, but that's not related to this PR.

Comment (srajangarg) [inline symengine/polynomial.cpp]: So, what should I stick with now?

Comment (isuruf) [inline symengine/polynomial.cpp]: For now, keep the above as it is in the master, we'll figure it out later.

**Comment by srajangarg** [comment]
@isuruf ping! Changes done from my side.

**Comment by isuruf** [comment]
@srajangarg, thanks for the PR
